### PR TITLE
Use MEF for all completion provider registrations

### PR DESCRIPTION
**Pull request by sharwell**
* Convert default completion providers to MEF registration, will full testing of deterministic order
* Remove 'exclusive providers' hook
* Remove 'SetTestProviders' hook

This pull request simplifies the completion provider logic as part of a move in the direction of #30270.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
LMK when you'd like a review.

**Comment by sharwell** [comment]
@CyrusNajmabadi ready!

**Comment by sandyarmstrong** [comment]
Thanks for the ping, Sam. This looks like something we'll need to test in VSmac before it gets merged here. I'll make time for that next week.

File: src/VisualStudio/Core/Test/Completion/VisualBasicMockCompletionProvider.vb [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+Imports System.Composition
+Imports Microsoft.CodeAnalysis
+Imports Microsoft.CodeAnalysis.Completion
+Imports Microsoft.CodeAnalysis.Host.Mef
+
+Namespace Microsoft.VisualStudio.LanguageServices.UnitTests.Completion
+    <ExportCompletionProvider(NameOf(VisualBasicMockCompletionProvider), LanguageNames.VisualBasic)>
+    <[Shared]>
+    <PartNotDiscoverable>
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Test/Completion/VisualBasicMockCompletionProvider.vb]: what's the general intuition around a non-discoverable part?  you can explicitly make a mef container that has it in it?  but if you ask for a mef container for an assembly, it won't be found/used automatically?

Comment (sharwell) [reply src/VisualStudio/Core/Test/Completion/VisualBasicMockCompletionProvider.vb]: > you can explicitly make a mef container that has it in it? but if you ask for a mef container for an assembly, it won't be found/used automatically?

Exactly 👍

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/CompletionProviderOrderTests.cs [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+                typeof(LastBuiltInCompletionProvider),
+            };
+
+            AssertEx.EqualOrDiff(
+                string.Join(Environment.NewLine, expectedOrder.Select(x => x.FullName)),
+                string.Join(Environment.NewLine, actualOrder.Select(x => x.FullName)));
+        }
+
+        [Fact]
+        public void TestCompletionProviderOrderMetadata()
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/CompletionProviderOrderTests.cs]: so we test the strict ordering above?

File: src/Features/CSharp/Portable/Completion/CompletionProviders/FirstBuiltInCompletionProvider.cs [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+using System.Composition;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Completion;
+
+namespace Microsoft.CodeAnalysis.CSharp.Completion.Providers
+{
+    [ExportCompletionProvider(nameof(FirstBuiltInCompletionProvider), LanguageNames.CSharp)]
+    [Shared]
+    internal sealed class FirstBuiltInCompletionProvider : CompletionProvider
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/CompletionProviders/FirstBuiltInCompletionProvider.cs]: probably worth documenting this.

File: src/Features/CSharp/Portable/Completion/CompletionProviders/LastBuiltInCompletionProvider.cs [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+using System.Composition;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Completion;
+
+namespace Microsoft.CodeAnalysis.CSharp.Completion.Providers
+{
+    [ExportCompletionProvider(nameof(LastBuiltInCompletionProvider), LanguageNames.CSharp)]
+    [ExtensionOrder(After = nameof(ExtensionMethodImportCompletionProvider))]
+    [Shared]
+    internal sealed class LastBuiltInCompletionProvider : CompletionProvider
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/CompletionProviders/LastBuiltInCompletionProvider.cs]: and this one too.

File: src/Workspaces/Remote/Core/Services/RoslynServices.cs [hunk 5]
```diff
@@ -32,13 +32,9 @@ internal sealed class RoslynServices
...
         public static readonly ImmutableArray<Assembly> RemoteHostAssemblies =
             MefHostServices.DefaultAssemblies
-                // This adds the exported MEF services from Workspaces.Desktop
-                .Add(typeof(Host.TemporaryStorageServiceFactory.TemporaryStorageService).Assembly)
                 // This adds the exported MEF services from the RemoteWorkspaces assembly.
                 .Add(typeof(RoslynServices).Assembly)
-                .Add(typeof(ICodingConventionsManager).Assembly)
-                .Add(typeof(CSharp.CodeLens.CSharpCodeLensDisplayInfoService).Assembly)
-                .Add(typeof(VisualBasic.CodeLens.VisualBasicDisplayInfoService).Assembly);
+                .Add(typeof(ICodingConventionsManager).Assembly);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/Core/Services/RoslynServices.cs]: nice!

File: src/EditorFeatures/TestUtilities/Workspaces/TestWorkspace_XmlConsumption.cs [hunk 6]
```diff
@@ -910,7 +910,13 @@ private static IList<MetadataReference> CreateReferenceList(TestWorkspace worksp
             var references = CreateCommonReferences(workspace, element);
             foreach (var reference in element.Elements(MetadataReferenceElementName))
             {
-                references.Add(MetadataReference.CreateFromFile(reference.Value));
+                // Read the image to an ImmutableArray<byte>, since the GC does a better job of tracking these and
+                // knowing when it's necessary to run finalizers to clean up old Metadata objects that are no longer in
+                // use. There are no public APIs available to directly dispose of these objects, so we are relying on GC
+                // running finalizers to avoid OutOfMemoryException during tests.
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/TestUtilities/Workspaces/TestWorkspace_XmlConsumption.cs]: 😱

**Comment by CyrusNajmabadi** [comment]
Overall, this looks totally good to me.  What are the top benefits we get from this?  I get the niceness of not having a mix of MEF discovery + static discovery.  Are there other benefits that are here as well?  Thanks!

File: src/EditorFeatures/TestUtilities/Workspaces/TestWorkspace_XmlConsumption.cs (hunk 6 above)
Comment (sharwell) [reply src/EditorFeatures/TestUtilities/Workspaces/TestWorkspace_XmlConsumption.cs]: 🤣

**Comment by jinujoseph** [comment]
blocked till VSMac validation is done

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/CompletionProviderOrderTests.cs (hunk 2 above)
Comment (sharwell) [reply src/EditorFeatures/CSharpTest/Completion/CompletionProviders/CompletionProviderOrderTests.cs]: We test the complete order above, but the sorting logic is based on a partial order. This method tests for deterministic order.

File: src/Workspaces/Remote/Core/Services/RoslynServices.cs (hunk 5 above)
Comment (sharwell) [reply src/Workspaces/Remote/Core/Services/RoslynServices.cs]: This was actually a bug (duplicate of the features assemblies). Looks like we never had any exports in them before.

**Comment by Therzok** [comment]
cc @sandyarmstrong

**Comment by sharwell** [comment]
ping @sandyarmstrong 😄

**Comment by sharwell** [comment]
ping @sandyarmstrong

**Comment by sandyarmstrong** [comment]
@sharwell Sorry for holding this up. Just got back from a vacation. I'll work on testing it today.

**Comment by kzu** [comment]
@sandyarmstrong unblocking this would mean that sooner than later we'd be able to really turn https://github.com/xamarin/CodeAnalysis into a cross-IDE 100% shared nuget package containing not only analysers and code fixes but also (quite key) completion providers. Otherwise, we need to also create separate VSIX and MPack and insert into the IDEs separately, which is quite a burden.

Is there any way we can help prioritize this?

Thanks!

**Comment by chamons** [comment]
Since the macios was working on an analyzer in CodeAnalysis, we have interest in moving this  forward so we can move that forward. :)

**Comment by sandyarmstrong** [comment]
@kzu sorry, we were very out-of-sync during the Roslyn 3.5 cycle and I ended up not having the bandwidth to test. We are much more in-sync for Roslyn 3.6 so I should be able to test when I'm back next Tuesday.

@sharwell would it be possible to merge latest master into this? Thanks!

**Comment by sandyarmstrong** [comment]
Working through build issues on Mac. Testing will continue into tomorrow.

**Comment by sandyarmstrong** [comment]
@sharwell I cannot find any behavior regressions in VSmac with this change.

We have a unit test that uses `CompletionServiceWithProviders.SetTestProviders`, which this PR removes, so I just have to see how to deal with that.

**Comment by sharwell** [comment]
@sandyarmstrong `SetTestProviders` is no longer necessary. It now uses the completion providers from the MEF catalog instead. If you want a specific set of providers for testing, just restrict your MEF catalog to only include those providers. 👍

**Comment by sharwell** [comment]
> Overall, this looks totally good to me. What are the top benefits we get from this? I get the niceness of not having a mix of MEF discovery + static discovery. Are there other benefits that are here as well?

It allows third-party completion providers to test their completion providers without the service including results from all the built-in providers.

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/CompletionProviderOrderTests.cs (hunk 2 above)
Comment (sharwell) [reply src/EditorFeatures/CSharpTest/Completion/CompletionProviders/CompletionProviderOrderTests.cs]: Added comments to the tests

File: src/Features/CSharp/Portable/Completion/CompletionProviders/FirstBuiltInCompletionProvider.cs (hunk 3 above)
Comment (sharwell) [reply src/Features/CSharp/Portable/Completion/CompletionProviders/FirstBuiltInCompletionProvider.cs]: Documented

File: src/Features/CSharp/Portable/Completion/CompletionProviders/LastBuiltInCompletionProvider.cs (hunk 4 above)
Comment (sharwell) [reply src/Features/CSharp/Portable/Completion/CompletionProviders/LastBuiltInCompletionProvider.cs]: Documented

File: src/Features/CSharp/Portable/Completion/CSharpCompletionService.cs [hunk 7]
```diff
@@ -33,58 +32,16 @@ public ILanguageService CreateLanguageService(HostLanguageServices languageServi
...
-
-            var languageServices = workspace.Services.GetLanguageServices(LanguageNames.CSharp);
-            var languagesProvider = languageServices.GetService<IEmbeddedLanguagesProvider>();
-            if (languagesProvider != null)
-            {
-                defaultCompletionProviders = defaultCompletionProviders.Add(
-                    new EmbeddedLanguageCompletionProvider(languagesProvider));
-            }
-
-            _defaultCompletionProviders = defaultCompletionProviders;
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/CSharpCompletionService.cs]: nice!

File: src/Features/CSharp/Portable/Completion/CompletionProviders/CrefCompletionProvider.cs [hunk 8]
```diff
@@ -422,5 +426,21 @@ private CompletionItemRules GetRules(string displayText)
...
+            => new TestAccessor(this);
+
+        internal readonly struct TestAccessor
+        {
+            private readonly CrefCompletionProvider _crefCompletionProvider;
+
+            public TestAccessor(CrefCompletionProvider crefCompletionProvider)
+            {
+                _crefCompletionProvider = crefCompletionProvider;
+            }
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/CompletionProviders/CrefCompletionProvider.cs]: nit:

```suggestion
             => _crefCompletionProvider = crefCompletionProvider;
```

i actually like that a lot for simple wrapper structs to make it clear this is just for exposing/hiding around a single value.

File: src/EditorFeatures/VisualBasicTest/Completion/CompletionProviderOrderTests.vb [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+
+        ''' <summary>
+        ''' Verifies that the order of built-in completion providers is deterministic.
+        ''' </summary>
+        <Fact>
+        Public Sub TestCompletionProviderOrderMetadata()
+            Dim completionProviderExports = TestExportProvider.ExportProviderWithCSharpAndVisualBasic.GetExports(Of CompletionProvider, CompletionProviderMetadata)()
+            Dim orderedVisualBasicCompletionProviders = ExtensionOrderer.Order(completionProviderExports.Where(Function(export) export.Metadata.Language = LanguageNames.VisualBasic))
+
+            For i = 0 To orderedVisualBasicCompletionProviders.Count - 1
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/VisualBasicTest/Completion/CompletionProviderOrderTests.vb]: would be nice to share test code for this.  but it's minor.

File: src/Features/CSharp/Portable/Completion/CompletionProviders/LastBuiltInCompletionProvider.cs [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+
+using System.Composition;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Completion;
+
+namespace Microsoft.CodeAnalysis.CSharp.Completion.Providers
+{
+    /// <summary>
+    /// Provides a completion provider that always appears after all built-in completion providers. This completion
+    /// provider does not provide any completions.
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/CompletionProviders/LastBuiltInCompletionProvider.cs]: what is this for then? is this just for testing purposes?

File: src/Features/Core/Portable/Completion/Providers/AbstractEmbeddedLanguageCompletionProvider.cs [hunk 11]
```diff
@@ -20,20 +22,36 @@ namespace Microsoft.CodeAnalysis.Completion.Providers
...
 
         public override bool ShouldTriggerCompletion(SourceText text, int caretPosition, CompletionTrigger trigger, OptionSet options)
         {
-            foreach (var language in _languageProviders)
+            Func<object, IEmbeddedLanguagesProvider> languagesProvider =
+                state =>
+                {
+                    var document = ((SourceText)state).GetOpenDocumentInCurrentContextWithChanges();
+                    return document.Project.LanguageServices.GetService<IEmbeddedLanguagesProvider>();
+                };
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/Providers/AbstractEmbeddedLanguageCompletionProvider.cs]: why do these need to be Funcs that take objects?  that part isn't really clear to me.

File: src/Features/CSharp/Portable/Completion/CompletionProviders/LastBuiltInCompletionProvider.cs (hunk 10 above)
Comment (sharwell) [reply src/Features/CSharp/Portable/Completion/CompletionProviders/LastBuiltInCompletionProvider.cs]: Completion providers are a public API. This allows users to provide a completion provider at the beginning or end of the order.

File: src/Features/Core/Portable/Completion/Providers/AbstractEmbeddedLanguageCompletionProvider.cs (hunk 11 above)
Comment (sharwell) [reply src/Features/Core/Portable/Completion/Providers/AbstractEmbeddedLanguageCompletionProvider.cs]: Fixed

**Comment by Thaina** [comment]
Is this already made `completion as a package` completely work in next version of roslyn or do we have anything else need to be waited for?

**Comment by sharwell** [comment]
@Thaina Support for completion providers in NuGet packages is a separate feature that is not yet implemented.

**Comment by Thaina** [comment]
@sharwell Thank you for your information. But then, how many things we need to do more than this? Do we have a roadmap or checklist of what we need for that feature?

Also, while this feature is not finished yet, are there any sample of how to do CompletionProvider registeration in vscode omnisharp ?

**Pull request merged**